IPIP/0000-cidv2-tagged-content-identifiers.md

```diff
@@ -0,0 +1,130 @@
+<!--One paragraph explanation of the IPIP.-->
+Create a new [CID](https://github.com/multiformats/cid) version (CIDv2, informally "Tagged Content Identifiers") which combines a data Multicodec-Multihash pair (the pointer) and a metadata Multicodec-Multihash pair (the tag) to create content-addresses with expressive contexts.
+
+## Motivation
+
+Currently, CIDv1 data is described by a multicodec content type. However, this is meant to describe the overall format of the serialized data e.g. the `dag-cbor` IPLD encoding, and not more specific information such as a data schema or type. For example, it can be useful to have raw IPLD data contextualized by its [IPLD schema](https://ipld.io/docs/schemas/intro/). Since multicodecs are limited to 9 bytes  by the [unsigned-varint spec](https://github.com/multiformats/unsigned-varint#practical-maximum-of-9-bytes-for-security), the available codec space is generally too small to encode such metadata.
+
+## Detailed design
+
+Our solution is a new CID version which contains two multicodec-multihash pairs, one pair for data and another for metadata. The metadata multicodec would be able to concisely describe a space of metadata tags where the specific tag would then be further specified by the multihash. This could be implemented as follows in Rust:
+
+```rust
+pub struct CidV2<const S: usize, const M: usize> {
+    /// The data multicodec
+    data_codec: u64,
+    /// The data multihash
+    data_hash: Multihash<S>,
+    /// The metadata codec
+    meta_codec: u64,
+    /// The metadata multihash
+    meta_hash: Multihash<M>
+}
+```
+
+It would serialize as follows:
+
+```
+<cidv2> ::= <multicodec-cidv2><multicodec-data-content-type><multihash-data><multicodec-metadata-content-type><multihash-metadata>
+```
+
+with a multibase prefix when represented in text.
+
+For example, suppose you want a CID which points to a piece of IPLD data and its [IPLD schema](https://ipld.io/docs/schemas/). Let's say you have the schema `Trit`, with a particular integer representation 
+
+```
+type Trit union {
+  | True ("1")
+  | False ("2")
+  | Unknown ("0")
+} representation int
+```
+
+which corresponds to the Ipld data: `Ipld::Num(1)`, `Ipld::Num(2)`, `Ipld::Num(0)`.
+
+While you could in principle propose a new multicodec for `Trit`, this might be not suitable if `Trit` is a temporary or ephemeral structure, or if you have a large number of different schemas (For instance, in Lurk-lang's content-addressing we would need to reserve 16-bits of the multicodec table, or 2^16 distinct multicodecs).
+
+However, since IPLD schemas can be [represented as JSON](https://ipld.io/specs/schemas/#dsl-vs-dmt) and hashed, with a CIDv2 we could reserve a single IPLD schema multicodec, along with the codec for the data representation (such as dag-cbor)
+We could then use the above CIDv2 definiton to create a pointer to any Schema+Data pair:
+
+```
+CidV2 { 
+  data_codec: 0x71,
+  data_hash: <data_multihash>,
+  meta_codec : 0x3e7ada7a, 
```

Comment:
> Having a single code for "IPLD schema" followed by a multihash seems off. IPLD schemas can be represented in multiple different formats including dag-json and dag-cbor. Is this a codec for ipld-schema-dag-json

There are a lot of things in multicodec for which this is also true (e.g. the ethereum codecs: https://github.com/multiformats/multicodec/blob/master/table.csv#L55) and my understanding of how it works there is that the format is just described in the description (such as https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/, even though you could in principle encode any RLP data as dag-cbor if you wanted)

Reply:
>  There are a lot of things in multicodec for which this is also true ... the ethereum codecs ... even though you could in principle encode any RLP data as dag-cbor if you wanted

IPLD codecs tell you how to decode serialized representations of data (into the IPLD data model),  not necessarily what the data is or what it's for. The ethereum codecs, like the Git ones are tied to a particular serialized data format if you wanted to transcode the data into something like dag-cbor tagging the data with the prior codec would result in a deserialization error.

Many existing hash linked data structures have more fixed representations then say the [FBL ADL](https://ipld.io/specs/advanced-data-layouts/fbl/) which is defined over arbitrary serialized forms as long as they can be decoded into a compatible IPLD Data Model layout. As a result it can appear as though the codecs are types even though they're deserialization mechanisms.

This means really what you'd need to express the <ipld-schema> type correctly is a second code to say "this is dag-json" next to the code saying it was an ipld-schema.

Reply:
> IPLD codecs tell you how to decode serialized representations of data (into the IPLD data model), not necessarily what the data is or what it's for.

The key word is *necessarily* there. A multicodec can absolutely tell you what the data is for. For example, if you have a CIDv1 that points to an Ethereum Block you could equally choose to encode using 

|name | tag| code | description |
|-----|----|-----|-|
|rlp|                            serialization|  0x60|   recursive length prefix |
| eth-block |                  ipld|           0x90 | Ethereum Header (RLP) |

Likewise we have a codec for all `cbor` and a more specific codec for `dag-cbor`. And ofc `raw` supersets everything.

So there's nothing strange if the IPLD Schema team wanted to set a default format 

|name | tag| code | description |
|-----|----|-----|-|
| ipld-schema |                  ipld|           0xdead_beef | Ipld Schema (dag-cbor) |

or with dag-json. Yes we could have `ipld-schema-dag-json`, `ipld-schema-dag-cbor` to disambiguate, but that seems like it should be an application level decision

Reply:
> The key word is necessarily there. A multicodec can absolutely tell you what the data is for.

If doing nominative typing this way was reasonable then CIDv2 wouldn't be necessary in almost any application since you could just register every type as a different codec. IIUC this kind of thing would in theory work for the UCAN case as well, it's just that using the global code table for nominative typing like this seems bad. Applications can end up with many different named data types, sometimes it's 10s or 100s, or the many more that Lurk would require reserving codes in the table this way.

Some links around not using multicodecs for nominatives types:
- https://github.com/ipld/ipld/blob/master/_legacy/specs/FOUNDATIONS.md#multicodecs-are-not-meant-to-act-as-types
- https://github.com/multiformats/multicodec/issues/204

> Yes we could have `ipld-schema-dag-json`, `ipld-schema-dag-cbor` to disambiguate, but that seems like it should be an application level decision

Sure, but how can I do a non-disambiguated `ipld-schema` that just works like IPLD Schemas do on any IPLD Data Model data? This code field has provided nominative typing, but without enough parameters to be useful for parameterized nominative types like IPLD Schemas (or dealing with multiblock data structures as in (https://github.com/ipfs/specs/pull/305#discussion_r940062425)

Reply:
I'm glad this discussion is happening here, although I feel it might be a diversion from the main discussion. This specific point is why I was hoping to have @vmx chime in. I worry that the Lurk specifics embedded in the doc here might be a distraction from the main goal. Even after reading all of this I don't really understand why, with the second CID-ish for metadata Lurk just couldn't encode a dag-cbor, dag-json, or even raw custom format bytes with the tag they want. Specifically: `meta_codec` could be dag-cbor (`0x71`), and `meta_hash` be inline (`0x0`) with whatever you like for your tag—you could even embed the mega-int here that the 9-byte varints are getting in the way of currently.
Perhaps that's essentially what you're aiming for through the use of a new "codec" to identify a "schema", just keeping it more efficient.

**But** my point again is that I think this is a distraction for the purpose of this spec. If Lurk wants to abuse the multicodec spec then that's their choice. It would be best for everyone if they want to register a new codec for this purpose to identify a "schema" in the multicodec table and we could continue this discussion there. For now, I think `0x3e7ada7a` is in the way. It stands apart from the commonly understood purpose of a CID and as this discussion is suggesting there's a weirdness about it that leads us into a deep hole (the multicodec repo has _many_ of these deep holes, covering very similar territory, I even had this discussion specifically about rlp and the eth codecs just a couple of months ago). We accept that there are squishy edges to the concept of a "multicodec", but always work to try and keep things toward the well understood and agreed-upon center _where possible_.

So my suggestion is to remove `0x3e7ada7a`, shunt this distraction to the multicodec repo in due course, and go with something more commonly understood - maybe just an inline dag-json blob. Then we can at least start reasoning about the basics.

Reply:
> But my point again is that I think this is a distraction for the purpose of this spec.

I agree. I see this just as an example of how people might want to use it. I think the purpose of the proposal should be about, whether we want those CIDs with two pointers to provide additional context or not.

In regards to Lurk, I also don't think the `0x3e7ada7a` codec is needed. It could just be e.g. DAG-CBOR and you could encode your schema as such.

Reply:
>  Even after reading all of this I don't really understand why, with the second CID-ish for metadata Lurk just couldn't encode a dag-cbor, dag-json, or even raw custom format bytes with the tag they want.

Just to clarify, we absolutely can, and this is part of the intent of having the second CID. The `0x3e7ada7a` was only meant as an illustrative example for `ipld-schemas`, but the rest of the proposal is the same if we just replaced it in the doc with `0x71` for `dag-cbor`, as @vmx suggested.

Regarding the prior topic about nominative types, I don't think either Lurk or Yatima need or want to add typing to multicodec, really. To be super concrete about what we need: For Lurk, we want to add a 16-bit metadata field to our CIDs, and for Yatima we want a 256-bit metadata multihash. In terms of multicodecs, I don't think it matters that much to either of those use cases whether we get a single application codec, multiple application codecs, or no codecs (and we just use e.g. `dag-cbor`). As long as we have a flexible way to add metadata, if we do end up needing additional info in pointers, we can just put it in that variable metadata field (e.g. with the identity multihash)